pkg/controller/v1alpha1/localmodelnode/controller.go

```diff
@@ -45,13 +53,209 @@ type LocalModelNodeReconciler struct {
+			Template: v1.PodTemplateSpec{
+				Spec: v1.PodSpec{
+					NodeName:      nodeName,
+					Containers:    []v1.Container{*container},
+					RestartPolicy: v1.RestartPolicyNever,
+					Volumes: []v1.Volume{
+						{
+							Name: "kserve-pvc-source",
+							VolumeSource: v1.VolumeSource{
+								PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+									ClaimName: claimName,
+								},
+							},
+						},
+					},
+					SecurityContext: &v1.PodSecurityContext{
+						FSGroup: FSGroup,
+					},
+				},
+			},
+		},
+	}
+	dryrunJob, err := jobs.Create(ctx, expectedJob, metav1.CreateOptions{DryRun: []string{metav1.DryRunAll}})
+	if err != nil {
+		return nil, err
+	}
+	if job != nil && reflect.DeepEqual(job.Spec.Template.Spec, dryrunJob.Spec.Template.Spec) {
+		return job, nil
+	}
+	if jobFound {
+		bg := metav1.DeletePropagationBackground
+		err = jobs.Delete(ctx, job.Name, metav1.DeleteOptions{
+			PropagationPolicy: &bg,
+		})
+		if err != nil {
+			c.Log.Error(err, "Failed to delete job.", "name", job.Name)
+			return nil, err
+		}
+	}
+
+	if err := controllerutil.SetControllerReference(localModelNode, expectedJob, c.Scheme); err != nil {
+		c.Log.Error(err, "Failed to set controller reference", "name", modelInfo.ModelName)
+		return nil, err
+	}
+	expectedJob.Name = jobName
+	job, err = jobs.Create(ctx, expectedJob, metav1.CreateOptions{})
+	c.Log.Info("Creating job", "name", job.Name, "namespace", job.Namespace)
+	if err != nil {
+		c.Log.Error(err, "Failed to create job.", "name", expectedJob.Name)
+		return nil, err
+	}
+	return job, err
+}
+
+// Fetches container spec for model download container, use the default KServe image if not found
+func (c *LocalModelNodeReconciler) getContainerSpecForStorageUri(ctx context.Context, storageUri string) (*v1.Container, error) {
+	storageContainers := &v1alpha1api.ClusterStorageContainerList{}
+	if err := c.Client.List(ctx, storageContainers); err != nil {
+		return nil, err
+	}
+
+	for _, sc := range storageContainers.Items {
+		if sc.IsDisabled() {
+			continue
+		}
+		if sc.Spec.WorkloadType != v1alpha1api.LocalModelDownloadJob {
+			continue
+		}
+		supported, err := sc.Spec.IsStorageUriSupported(storageUri)
+		if err != nil {
+			return nil, fmt.Errorf("error checking storage container %s: %w", sc.Name, err)
+		}
+		if supported {
+			return &sc.Spec.Container, nil
+		}
+	}
+
+	defaultContainer := &v1.Container{
+		Image:                    defaultJobImage,
+		TerminationMessagePolicy: v1.TerminationMessageFallbackToLogsOnError,
+	}
+	return defaultContainer, nil
+}
+
+func (c *LocalModelNodeReconciler) downloadModels(ctx context.Context, localModelNode *v1alpha1api.LocalModelNode) error {
+	c.Log.Info("Downloading models to", "node", localModelNode.ObjectMeta.Name)
+
+	for _, modelInfo := range localModelNode.Spec.LocalModels {
+		if status, ok := localModelNode.Status.ModelStatus[modelInfo.ModelName]; ok {
+			if status == v1alpha1api.ModelDownloaded {
```

Comment:
How do we handle the case if the model was downloaded but then node is rebuilt ?

Reply:
Going to handle it in a different PR.